ethereumj-core/src/main/java/org/ethereum/listener/RecommendedGasPriceTracker.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) [2016] [ <ether.camp> ]
+ * This file is part of the ethereumJ library.
+ *
+ * The ethereumJ library is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The ethereumJ library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with the ethereumJ library. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.ethereum.listener;
+
+import org.apache.commons.collections4.queue.CircularFifoQueue;
+import org.ethereum.core.Block;
+import org.ethereum.core.BlockSummary;
+import org.ethereum.core.Transaction;
+import org.ethereum.util.ByteUtil;
+
+import java.lang.reflect.Array;
+import java.util.Arrays;
+import java.util.List;
+
+
+/**
+ * Calculates a 'reasonable' Gas price based on statistics of the latest transaction's Gas prices.
+ * This is an updated version that returns more accurate data
+ * in networks with large number of transactions like Ethereum MainNet.
+ * However it needs more CPU and memory resources for processing.
+ *
+ * Normally the price returned should be sufficient to execute a transaction since ~25%
+ * (if {@link #getPercentileShare()} is not overridden) of the latest transactions were
+ * executed at this or lower price.
+ */
+public class RecommendedGasPriceTracker extends EthereumListenerAdapter {
+
+    private static final Long DEFAULT_PRICE = null;
+    private static final int MIN_BLOCKS = 128;
+    private static final int BLOCKS_RECOUNT = 1;
+    private static final int MIN_TRANSACTIONS = 512;
+    private static final int PERCENTILE_SHARE = 4;
+
+    private CircularFifoQueue<long[]> blockGasPrices;
+
+    private int idx = 0;
+    private Long recommendedGasPrice = getDefaultPrice();
+
+    public RecommendedGasPriceTracker() {
+        blockGasPrices = new CircularFifoQueue<>(Math.max(getMinTransactions(), getMinBlocks()));
```

Comment:
Since `gasPrices` collection became not responsible for purging old Txes `CircularFifoQueue` starts to be redundant. Maybe it's better to use `LinkedList` here?